lib/SimpleSAML/Metadata/MetaDataStorageHandlerMDX.php

```diff
@@ -0,0 +1,287 @@
+		case 'shib13-sp-remote':
+			return $entity->getMetadata1xSP();
+		case 'attributeauthority-remote':
+			return $entity->getAttributeAuthorities();
+
+		default:
+			SimpleSAML_Logger::warning('MetaData - Handler.MDX: Unknown metadata set: ' . $set);
+		}
+
+		return NULL;
+	}
+
+
+	/**
+	 * Overriding this function from the superclass SimpleSAML_Metadata_MetaDataStorageSource.
+	 *
+	 * This function retrieves metadata for the given entity id in the given set of metadata.
+	 * It will return NULL if it is unable to locate the metadata.
+	 *
+	 * This class implements this function using the getMetadataSet-function. A subclass should
+	 * override this function if it doesn't implement the getMetadataSet function, or if the
+	 * implementation of getMetadataSet is slow.
+	 *
+	 * @param $index  The entityId or metaindex we are looking up.
+	 * @param $set  The set we are looking for metadata in.
+	 * @return An associative array with metadata for the given entity, or NULL if we are unable to
+	 *         locate the entity.
+	 */
+	public function getMetaData($index, $set) {
+		assert('is_string($index)');
+		assert('is_string($set)');
+
+		SimpleSAML_Logger::info('MetaData - Handler.MDX: Loading metadata entity [' . $index . '] from [' . $set . ']' );
+
+		/* Read from cache if possible. */
+		$data = $this->getFromCache($set, $index);
+
+		if ($data !== NULL && array_key_exists('expires', $data) && $data['expires'] < time()) {
+			/* Metadata has expired. */
+			$data = NULL;
+		}
+
+		if (isset($data)) {
+			/* Metadata found in cache and not expired. */
+			SimpleSAML_Logger::debug('MetaData - Handler.MDX: Using cached metadata for: ' . $index . '.');
+			return $data;
+		}
+
+		/* Look at Metadata Query Protocol: https://github.com/iay/md-query/blob/master/draft-young-md-query.txt */
+		$mdx_url = $this->server . '/entities/' . urlencode( $index);
+
+		SimpleSAML_Logger::debug('MetaData - Handler.MDX: Downloading metadata for "'. $index .'" from [' . $mdx_url . ']' );
+		try {
+			$xmldata = SimpleSAML_Utilities::fetch($mdx_url);
+		} catch(Exception $e) {
+			SimpleSAML_Logger::warning('Fetching metadata for ' . $index . ': ' . $e->getMessage());
+		}
+
+		if (empty($xmldata)) {
+			throw new Exception('Error downloading metadata for "'. $index .'" from "' . $mdx_url . '": ' .
+				SimpleSAML_Utilities::getLastError());
+		}
+
+		$entity = SimpleSAML_Metadata_SAMLParser::parseString($xmldata);
+		SimpleSAML_Logger::debug('MetaData - Handler.MDX: Completed parsing of [' .
+			$mdx_url . ']'  );
+
+
+		if( $this->validateFingerprint !== NULL) {
+			if(!$entity->validateFingerprint($this->validateFingerprint)) {
+				throw new Exception ('Error, could not verify signature for entity: ' . $index . '".');
+			}
+		}
+
+		$data = self::getParsedSet($entity, $set);
+		if ($data === NULL) {
+			throw new Exception('No metadata for set "' . $set .
+				'" available from "' . $index . '".');
+		}
+		if (is_array($data[0])) $data = $data[0];
```

Comment:
I understand this is to handle the case when multiple entities are returned from the MDX server, which I don't know if it's even possible. In any case, I assume the most common case would be the MDX server will return exactly one entity when queried for an `entityID`, and in that case this code generates a warning in the log file, since the index 0 won't exist in the parsed metadata array. So if we want to see if the result from `getParsedSet()` returns a bunch of entities, we should do something like this:

```php
        $multiple_entities = true;
        foreach (array_keys($data) as $key) {
            if (!is_numeric($key)) {
                $multiple_entities = false;
                break;
            }
            if (!is_array($data[$key])) {
                $multiple_entities = false;
                break;
            }
        }
        if ($multiple_entities) {
            $data = reset($data);
        }
```

Basically, we assume we have multiple entities, and then we iterate over the keys of the array. As soon as we find one key that's not numeric or is not an array, we know this is not an array with multiple entities. If the loop finishes correctly, then we can fetch the first entity because it's probable that we got a bunch of them.

Could you please fix this in the pull request? As soon as this is fixed I think the code can be merged :+1: